gloo/examples/example_allreduce.cc

```diff
@@ -0,0 +1,187 @@
+  // multi-homed machines that all share the same network interface
+  // name, for example.
+  //
+  gloo::transport::uv::attr attr;
+  //attr.iface = "eth0";
+  //attr.iface = "ib0";
+  //attr.iface = "Wi-Fi";
+
+  // attr.ai_family = AF_INET; // Force IPv4
+  // attr.ai_family = AF_INET6; // Force IPv6
+  attr.ai_family = AF_UNSPEC; // Use either (default)
+
+  // A string is implicitly converted to an "attr" struct with its
+  // hostname field populated. This will try to resolve the interface
+  // to use by resolving the hostname or IP address, and finding the
+  // corresponding network interface.
+  //
+  // Hostname "localhost" should resolve to 127.0.0.1, so using this
+  // implies that all connections will be local. This can be useful
+  // for single machine operation.
+  //
+  //   auto dev = gloo::transport::tcp::CreateDevice("localhost");
+  //
+  std::cout << "CreateDevice" << std::endl;
+  auto dev = gloo::transport::uv::CreateDevice(attr);
+  // Now that we have a device, we can connect all participating
+  // processes. We call this process "rendezvous". It can be performed
+  // using a shared filesystem, a Redis instance, or something else by
+  // extending it yourself.
+  //
+  // See "gloo/rendezvous/store.h" for the functionality you need to
+  // implement to create your own store for performing rendezvous.
+  //
+  // Below, we instantiate rendezvous using the filesystem, given that
+  // this example uses multiple processes on a single machine.
+  //
+  auto fileStore = gloo::rendezvous::FileStore("/libtmp");
+  std::cout << "FileStore" << std::endl;
+
+  // To be able to reuse the same store over and over again and not have
+  // interference between runs, we scope it to a unique prefix with the
+  // PrefixStore. This wraps another store and prefixes every key before
+  // forwarding the call to the underlying store.
+  std::string prefix = getenv("PREFIX");
+  auto prefixStore = gloo::rendezvous::PrefixStore(prefix, fileStore);
+  std::cout << "prefixStore" << std::endl;
+
+  // Using this store, we can now create a Gloo context. The context
+  // holds a reference to every communication pair involving this
+  // process. It is used by every collective algorithm to find the
+  // current process's rank in the collective, the collective size,
+  // and setup of send/receive buffer pairs.
+  const int rank = atoi(getenv("RANK"));
+  const int size = atoi(getenv("SIZE"));
+  auto context = std::make_shared<gloo::rendezvous::Context>(rank, size);
+  try{
+    context->connectFullMesh(prefixStore, dev);
+    std::cout << "connectFullMesh" << std::endl;
+    // All connections are now established. We can now initialize some
+    // test data, instantiate the collective algorithm, and run it.
+    std::array<int, 4> data;
+    std::cout << "Input: " << std::endl;
+    for (int i = 0; i < data.size(); i++) {
+      data[i] = i;
+      std::cout << "data[" << i << "] = " << data[i] << std::endl;
```

Comment:
these are debug vestige?

Reply:
yes, for debugging purpose, will remove them all.